agent/csi/volumes.go

```diff
@@ -0,0 +1,263 @@
+package csi
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"sync"
+	"time"
+
+	"github.com/docker/swarmkit/agent/exec"
+	"github.com/docker/swarmkit/api"
+	"github.com/docker/swarmkit/log"
+)
+
+// volumes is a map that keeps all the currently available volumes to the agent
+// mapped by volume ID.
+type volumes struct {
+	mu     sync.RWMutex // To sync between Add(), Remove() and Reset() for "m" map
+	lock   sync.RWMutex // To sync between Get(), addVolumes() and removeVolumes()
+	m      map[string]*api.VolumeAssignment
+	plugin *NodePlugin
+}
+
+const maxRetry float64 = 25
+
+const initialBackoff = 100 * time.Millisecond
+const factor float64 = 2
+
+// NewManager returns a place to store volumes.
+func NewManager() exec.VolumesManager {
+	return &volumes{
+		m:      make(map[string]*api.VolumeAssignment),
+		plugin: NewNodePlugin(),
```

Comment:
This needs to be a collection of some type, probably a map, because the node can have many possible plugins. You need to look at the VolumeAssignment to know which plugin to use.